hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeysRenameRequest.java

```diff
@@ -0,0 +1,304 @@
+import static org.apache.hadoop.ozone.OzoneConsts.RENAMED_KEYS_MAP;
+import static org.apache.hadoop.ozone.OzoneConsts.UNRENAMED_KEYS_MAP;
+import static org.apache.hadoop.ozone.OzoneConsts.VOLUME;
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.BUCKET_LOCK;
+
+/**
+ * Handles rename keys request.
+ */
+public class OMKeysRenameRequest extends OMKeyRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMKeysRenameRequest.class);
+  private String volumeName = null;
+  private String bucketName = null;
+
+  public OMKeysRenameRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+
+    RenameKeysRequest renameKeys = getOmRequest().getRenameKeysRequest();
+    Preconditions.checkNotNull(renameKeys);
+
+    List<RenameKeyRequest> renameKeyList = new ArrayList<>();
+    for (RenameKeyRequest renameKey : renameKeys.getRenameKeyRequestList()) {
+      // Set modification time in preExecute.
+      KeyArgs.Builder newKeyArgs = renameKey.getKeyArgs().toBuilder()
+          .setModificationTime(Time.now());
+      renameKey.toBuilder().setKeyArgs(newKeyArgs);
+      renameKeyList.add(renameKey);
+
+    }
+    RenameKeysRequest renameKeysRequest = RenameKeysRequest
+        .newBuilder().addAllRenameKeyRequest(renameKeyList).build();
+    return getOmRequest().toBuilder().setRenameKeysRequest(renameKeysRequest)
+        .setUserInfo(getUserInfo()).build();
+  }
+
+  @Override
+  @SuppressWarnings("methodlength")
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+
+    RenameKeysRequest renameKeysRequest = getOmRequest().getRenameKeysRequest();
+    OMClientResponse omClientResponse = null;
+    // fromKeyName -> toKeyName
+    List<RenameKeyArgs> unRenamedKeys = new ArrayList<>();
+
+    List<OmRenameKeyInfo> renameKeyInfoList = new ArrayList<>();
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumKeyRenames();
+
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+        getOmRequest());
+
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    IOException exception = null;
+    OmKeyInfo fromKeyValue = null;
+
+    Result result = null;
+    Map<String, String> auditMap = new LinkedHashMap<>();
+    String fromKeyName = null;
+    String toKeyName = null;
+    boolean acquiredLock = false;
+    boolean renameStatus = true;
+
+    try {
+      for (RenameKeyRequest renameKeyRequest : renameKeysRequest
+          .getRenameKeyRequestList()) {
+        OzoneManagerProtocolProtos.KeyArgs keyArgs =
+            renameKeyRequest.getKeyArgs();
+
+        // resolve Bucket Link at the first time.
+        if (bucketName == null) {
```

Comment:
For all other keys I think we should verify that the volume/bucket is the same as for the first key.